### ISSUE TITLE
get order in per group error

### ISSUE THREAD

**Issue by 466300750** [pingcap/tidb#22133]
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

1. create table progress_td and insert test data

```
   create table progress_td
(
    date             timestamp     null,
    brand_id         bigint        null,
    brand_name       tinytext      null,
    stock_group_id   bigint        null,
    stock_group_name mediumtext    null,
    progress_ytd     decimal(6, 4) null
);

INSERT INTO progress_td (date, brand_id, brand_name, stock_group_id, stock_group_name, progress_ytd)
VALUES ('2020-12-30 00:00:00', 1, '品牌1', 1, '股份集团1', 0.6000),
       ('2020-12-30 00:00:00', 2, '品牌2', 1, '股份集团1', 0.5000),
       ('2020-12-30 00:00:00', 6, '品牌6', 3, '股份集团3', 0.0200),
       ('2020-12-30 00:00:00', 5, '品牌5', 3, '股份集团3', 0.0250),
       ('2020-12-30 00:00:00', 4, '品牌4', 2, '股份集团2', 0.0333),
       ('2020-12-30 00:00:00', 3, '品牌3', 2, '股份集团2', 0.1000),
       ('2020-12-30 00:00:00', 7, '品牌7', 3, '股份集团3', 0.0167);
```

2. execute select sql

```
(select brand_id,
        brand_name,
        @rank_in_sg := IF(@curSG = t.stock_group_id, @rank_in_sg + 1, 1) AS progress_ytd_rank_in_sg,
        @curSG := stock_group_id                                         as stock_group_id
 from progress_td t,
      (select @rank_in_sg := 0) x,
      (select @curSG := '') y
 order by t.stock_group_id, t.progress_ytd desc);
```

### 2. What did you expect to see? (Required)
brand_id|brand_name|progress_ytd_rank_in_sg|stock_group_id
--------|-------------|------------------------|---------------
1|	品牌1|	1|	1
2|	品牌2|	2|	1
3|	品牌3|	1|	2
4|	品牌4|	2|	2
5|	品牌5|	1|	3
6|	品牌6|	2|	3
7|	品牌7|	3|	3

### 3. What did you see instead (Required)
brand_id|brand_name|progress_ytd_rank_in_sg|stock_group_id
--------|-------------|------------------------|---------------
1|	品牌1|	1|	1
2|	品牌2|	2|	1
3|	品牌3|	2|	2
4|	品牌4|	1|	2
5|	品牌5|	2|	3
6|	品牌6|	1|	3
7|	品牌7|	1|	3

### 4. What is your TiDB version? (Required)

Release Version: v4.0.0-beta.2-1943-g857e3da6e
Edition: Community
Git Commit Hash: 857e3da6ebc17d1c17e4f150fa21c6c474456f1a
Git Branch: master
UTC Build Time: 2020-12-31 07:03:16
GoVersion: go1.15.2
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false

**Comment by morgo** [comment]
Here is a reduced testcase. I can verify this exists against master:
```sql
drop table if exists t;
create table t (
 stock_group_id bigint null,
 progress_ytd decimal(6, 4) null
);

INSERT INTO t (stock_group_id, progress_ytd)
VALUES
 (1, 0.6000),
 (1, 0.5000),
 (3, 0.0200),
 (3, 0.0250),
 (2, 0.0333),
 (2, 0.1000),
 (3, 0.0167);

-- differs from MySQL:
SELECT
 @rank_in_sg := IF(@curSG = t.stock_group_id, @rank_in_sg + 1, 1) AS progress_ytd_rank_in_sg,
 @curSG := stock_group_id as stock_group_id
 FROM
  t,
  (select @rank_in_sg := 0) x,
  (select @curSG := '') y
ORDER BY t.stock_group_id, t.progress_ytd DESC;
```

The problem appears to be caused by TiDB not reading the rows as they appear in `ORDER BY`:

```sql
-- correct/same as MySQL:
SELECT
 @rank_in_sg := IF(@curSG = t.stock_group_id, @rank_in_sg + 1, 1) AS progress_ytd_rank_in_sg,
 @curSG := stock_group_id as stock_group_id
 FROM
  t,
  (select @rank_in_sg := 0) x,
  (select @curSG := '') y;
```

I thought that this might be an incorrect optimization because the columns in the order by are not in the `SELECT` list, but this appears to make no difference. The bug still shows up in these two queries:

```sql
-- wrong result:
SELECT
 t.stock_group_id as t_stock_group_id,
 t.progress_ytd as t_progress_ytd,
 @rank_in_sg := IF(@curSG = t.stock_group_id, @rank_in_sg + 1, 1) AS progress_ytd_rank_in_sg,
 @curSG := stock_group_id as stock_group_id
 FROM
  t,
  (select @rank_in_sg := 0) x,
  (select @curSG := '') y
ORDER BY t.stock_group_id, t.progress_ytd DESC;

-- wrong result:
SELECT
 t.stock_group_id as t_stock_group_id,
 t.progress_ytd as t_progress_ytd,
 @rank_in_sg := IF(@curSG = t.stock_group_id, @rank_in_sg + 1, 1) AS progress_ytd_rank_in_sg,
 @curSG := stock_group_id as stock_group_id
 FROM
  t,
  (select @rank_in_sg := 0) x,
  (select @curSG := '') y
ORDER BY t_stock_group_id, t_progress_ytd DESC;
```

**Comment by XuHuaiyu** [comment]
- This is **NOT** a bug, I'll explain why later.
- This SQL uses `get_var` and `set_var` to mock the windows function `row_number()`. This is not recommended. 
- I suggest to use `select brand_id, brand_name, row_number() over (partition by stock_group_id order by progress_ytd ) from progress_td;` instead.

**Comment by lzmhhh123** [comment]
You can refer to the docs here  https://dev.mysql.com/doc/refman/8.0/en/user-variables.html

```
HAVING, GROUP BY, and ORDER BY, when referring to a variable that is assigned a value in the select expression list do not work as expected because the expression is evaluated on the client and thus can use stale column values from a previous row.

User variables are intended to provide data values. They cannot be used directly in an SQL statement as an identifier or as part of an identifier, such as in contexts where a table or database name is expected, or as a reserved word such as SELECT. This is true even if the variable is quoted.
```

**Comment by lzmhhh123** [comment]
@466300750 I suggest you using the window function instead of user variables.

**Comment by 466300750** [comment]
OK, thanks very much. The reason why I did not use window_function is that I tested `select brand_id, brand_name, row_number() over (partition by stock_group_id order by progress_ytd ) from progress_td;` in my local mysql docker container before(the almost same version and innodb_version with my used tidb). But this feature was not supported in that version. So I did not test this feature in my tidb. Thanks for your response very much.

**Comment by 466300750** [comment]
OK, thanks very much. The reason why I did not use window_function is that I tested `select brand_id, brand_name, row_number() over (partition by stock_group_id order by progress_ytd ) from progress_td;` in my local mysql docker container before(the almost same version and innodb_version with my used tidb). But this feature was not supported in that version. So I did not test this feature in my tidb. Thanks for your response very much.



> @466300750 I suggest you using the window function instead of user variables.
